csharp/EndpointDiscoveryLib/ScheduledEventsUtility.cs

```diff
@@ -0,0 +1,140 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+*/
+namespace ScheduledEventsEndpointDiscovery
+{
+    using System;
+    using System.Net;
+    using System.Net.NetworkInformation;
+    using System.Threading;
+
+    public static class ScheduledEventsUtility
+    {
+        public static readonly int InstanceMetadataServicePort = 8080;
+        private static readonly string InstanceMetadataServiceEndpointPath = "/metadata/latest/scheduledevents";
+        private const uint ControlSystemDhcpOptionID = 245;
+
+        /// <summary>
+        /// Use this method to get the URI for scheduled events
+        /// </summary>
+        /// <returns></returns>
+        public static Uri ObtainScheduledEventsUri()
+        {
+            IPAddress address = GetDhcpAddress();
+
+            if (address != null)
+            {
+                Uri endpoint = ConstructScheduledEventsEndpoint(address);
+                return endpoint;
+            }
+
+            return null;
+        }
+
+        // This implementation is copied from the Windows GuestAgent implementation
+        private static IPAddress GetDhcpAddress()
+        {
+            foreach (NetworkInterface nic in DhcpClient.GetDhcpInterfaces())
+            {
+                if (nic.OperationalStatus != OperationalStatus.Up)
+                {
+                    continue;
+                }
+
+                try
+                {
+                    IPAddress address = RequestControlSystemEndpointAddress("WindowsAzureGuestAgent", nic);
+                    return address;
+                }
+                catch (DhcpRequestFailedException e)
+                {
+                    
+                }
+            }
+
+            return null;
+        }
+
+        internal static Uri ConstructScheduledEventsEndpoint(IPAddress address)
+        {
+            if (address == null)
+            {
+                throw new ArgumentNullException(nameof(address));
+            }
+
+            UriBuilder builder = new UriBuilder();
+            builder.Scheme = "http";
+            builder.Host = address.ToString();
+            builder.Port = InstanceMetadataServicePort;
+            builder.Path = InstanceMetadataServiceEndpointPath;
+            return builder.Uri;
+        }
+
+        private static IPAddress RequestControlSystemEndpointAddress(string appID, NetworkInterface nic)
+        {
+            IPAddress address = null;
+            DhcpRequestFailedException requestException = null;
+
+            var thread = new Thread(() =>
```

Comment:
Gonna hold off on refactoring code pulled from GA code for now to get this out quicker